File: baekjoon/baekjoon-1991.cpp
#include <iostream>
using namespace std;

// 아니 왜 트리로 구현하면 메모리 에러나... 이진트리 공부하려고 푼 문젠데..
struct Node
{
  char left;
  char right;
};

struct Node arr[26];

void preorder(char root)
{
  if (root == '.')
    return;
  cout << root;
  preorder(arr[root].left);
  preorder(arr[root].right);
}

void inorder(char root)
{
  if (root == '.')
    return;

  inorder(arr[root].left);
  cout << root;
  inorder(arr[root].right);
}

void postorder(char root)
{
  if (root == '.')
    return;

  postorder(arr[root].left);
  postorder(arr[root].right);
  cout << root;
}

int main()
{
  int n;
  char p1, p2, p3;
  cin >> n;

  for (int i = 0; i < n; i++)
  {
    cin >> p1 >> p2 >> p3;
    arr[p1].left = p2;
    arr[p1].right = p3;
  }

  preorder('A');
  cout << endl;
  inorder('A');
  cout << endl;
  postorder('A');
  cout << endl;
}

File: baekjoon/baekjoon-1991.py
tree = {}

def preorder(x):
    if x != '.':
        print(x, end='')

        if x in tree:
            preorder(tree[x][0])
            preorder(tree[x][1])

def inorder(x):
    if x != '.':
        if x in tree:
            inorder(tree[x][0])
        
        print(x, end='')

        if x in tree:
            inorder(tree[x][1])

def postorder(x):
    if x != '.':
        if x in tree:
            postorder(tree[x][0])
            postorder(tree[x][1])
            
        print(x, end='')


if __name__ == '__main__':
    n = int(input())
    temp = []

    for i in range(n):
        root, left, right = map(str, input().split())
        tree[root] = [left, right]

    preorder('A')
    print()
    inorder('A')
    print()
    postorder('A')
